src/root.tsx

```diff
@@ -332,13 +332,15 @@ function isVNode(maybeNode: ComponentChild): maybeNode is VNode<unknown> {
   );
 }
 
-function childrenToTree(children: ComponentChild, root: Root<any>) {
-  return array(children).reduce(
-    (accumulator: {children: NodeTree, descendants: NodeTree}, next: ComponentChild) => {
-      accumulator.children.push(buildElementWrappers(next, root)[0]);
-      accumulator.descendants.push(...buildElementWrappers(next, root));
-      return accumulator;
-    },
-    {children: [] as NodeTree, descendants: [] as NodeTree},
-  );
+function childrenToTree(inputChildren: ComponentChild, root: Root<any>) {
+  const children: NodeTree = [];
+  const descendants: NodeTree = [];
+  
+  for (const child of array(inputChildren)) {
+    const wrappers = buildElementWrappers(child, root);
```

Comment:
This is the place where we recurse (`buildElementWrappers` in turn calls `childrenToTree`). Previously we did it twice per child which would then do it twice per child... you get the idea.